src/gotcha_dl.c

```diff
@@ -5,7 +5,124 @@
 #include "elf_ops.h"
 #include <dlfcn.h>
 
+#if defined(__GLIBC__) && __GLIBC__ >= 2 && __GLIBC_MINOR__ < 34
 void* _dl_sym(void* handle, const char* name, void* where);
+#else
+/**
+ * structure used to pass lookup addr and return library address.
+ */
+struct Addrs {
+    ElfW(Addr) library_laddr;
+    ElfW(Addr) lookup_addr;
+};
+/**
+ * This is a callback to get headers for each library.
+ * We check if the caller's virtual address is between base address and the virtual addr + memory size.
+ *
+ * @param info, information about headers
+ * @param size, size of headers
+ * @param data, data from caller
+ * @return
+ */
+int lib_header_callback(struct dl_phdr_info * info, size_t size, void * data) {
+    struct Addrs* addrs = data;
+    ElfW(Addr) end_addr = 0;
+    for (int i=0;i<info->dlpi_phnum;++i) {
+        ElfW(Addr) size = info->dlpi_phdr[i].p_vaddr + info->dlpi_phdr[i].p_memsz;
+        if (end_addr < size) end_addr = size;
+    }
+    ElfW(Addr) base_addr = info->dlpi_addr;
+    ElfW(Addr) end_phdr = base_addr + end_addr;
+    if (base_addr <= addrs->lookup_addr && end_phdr >= addrs->lookup_addr)
+        addrs->library_laddr = base_addr;
+    return 0;
+}
+/**
+ * This function looks for all symbols defined in rela or rel.
+ *
+ * @param name, symbol name
+ * @param lmap, library where we search for symbol.
+ * @return symbol pointer
+ */
+int dl_lookup_symbol_x(const char *name, struct link_map *lmap, void** symbol) {
```

Comment:
Few things here. First, there's some code in gotcha.c in the prepare_symbol() function that almost does what you need. That code iterates over every link map, checks if a symbol name is exported, then makes a certain call on any matching symbol. I'd suggest refactoring that code so it can server both prepare_symbol() and this use case (you won't want to iterate over ever link map, and you don't want the setInternalBindingAddressPointer call on the symbol, but the rest is a match).

What the below code is doing is a linear iteration over both exported symbol and imported symbols. That's going to get incorrect matches when it hits the imported symbols. Also, for fast look up of exported symbols there is a hash map in the binary (try googling 'elf hash' and/or 'gnu elf hash' for details).  That's what the prepare_symbols() code uses.

And if you ever do need to legitimately write code like the below, there is a FOR_EACH_PLTREL macro in elf_ops.h, which you can use to avoid writing duplicate code for both rel and rela.

Reply:
Ok i looked up the ```prepare_symbol``` code and used the ```lookup_gnu_hash_symbol``` part of code. In here I had to use the correct offset based on is_rela.